### ISSUE TITLE
ZLib gzip/gunzip implementation

### ISSUE THREAD

**Issue by jeff-kilbride** [dial-once/node-cache-manager-redis#38]
I am using your plugin, so thank you for the code!

I have forked the project and added compression of values as a config option to my version, using the built-in ZLib gzip/gunzip methods. The values I'm storing aren't huge, typically 1-10K, but I'm seeing them compressed to ~25% of their original size using the lowest (fastest) compression setting.

Is this something you would be interested in having as a contribution? If so, I will polish it up and try to put together some tests, before submitting a PR.

**Comment by mrister** [comment]
@jeff-kilbride sure! Please submit a PR. Thanks for a contribution.

**Comment by jeff-kilbride** [comment]
Ok, will do! Probably be a few days.

Something to consider...  as I have it, I'm not allowing any configuration of the ZLib instance. I just set it to the lowest / fastest compression and leave all the other values at their defaults. I don't want to complicate your code any more than necessary, but I'm wondering if other people will want to customize those settings. Do you think I should allow for more configuration or just keep it as simple as possible?

**Comment by mrister** [comment]
@jeff-kilbride  Since it is hard to predict all scenarios for people using the lib it would be nice if you can provide various options and if you have the capacity to do it. It's like redis options and should only be passed down to the needed instance with minimal code change to this lib.

**Comment by mrister** [comment]
@jeff-kilbride We might go for a release soon, any chance you can submit this PR soon so we can check it out?

**Comment by jeff-kilbride** [comment]
@mrister  Sorry for the delay. I have the basics working, I just haven't added the extra configuration, yet. I will take a look at it today. I also just realized that I've been making my commits to `develop`, so I'm going to fork the project again and put my stuff in a feature branch.

**Comment by jeff-kilbride** [comment]
@mrister  I just forked a fresh copy of the repo and the tests are no longer working for me. When I run `npm test`, this is what I get:

```
$ npm test

> cache-manager-redis@0.2.3 test /Users/jeff/WebstormProjects/node-cache-manager-redis
> jasmine

node_redis: Warning: Redis server does not require a password, but a password was supplied.
node_redis: Deprecated: The SETEX command contains a "undefined" argument.
This is converted to a "undefined" string now and will return an error from v.3.0 on.
Please handle this in your code to make sure everything works as you intended it to.
node_redis: Deprecated: The SETEX command contains a "undefined" argument.
This is converted to a "undefined" string now and will return an error from v.3.0 on.
Please handle this in your code to make sure everything works as you intended it to.
node_redis: Warning: Redis server does not require a password, but a password was supplied.
```
Notice there is no test output. When I run the tests in my old fork, they work as expected:
```
$ npm test

> cache-manager-redis@0.2.3 test /Users/jeff/WebstormProjects/node-cache-manager-redis.old
> jasmine

Started
....node_redis: Deprecated: The SETEX command contains a "undefined" argument.
This is converted to a "undefined" string now and will return an error from v.3.0 on.
Please handle this in your code to make sure everything works as you intended it to.
.node_redis: Deprecated: The SETEX command contains a "undefined" argument.
This is converted to a "undefined" string now and will return an error from v.3.0 on.
Please handle this in your code to make sure everything works as you intended it to.
.............................


34 specs, 0 failures
Finished in 0.153 seconds
```
Has something changed recently that would affect this? I tried removing the `auth_pass` line from `spec/config.json`, but it just gave me a different error:
```
$ npm test

> cache-manager-redis@0.2.3 test /Users/jeff/WebstormProjects/node-cache-manager-redis
> jasmine

node_redis: Deprecated: The SETEX command contains a "undefined" argument.
This is converted to a "undefined" string now and will return an error from v.3.0 on.
Please handle this in your code to make sure everything works as you intended it to.
node_redis: Deprecated: The SETEX command contains a "undefined" argument.
This is converted to a "undefined" string now and will return an error from v.3.0 on.
Please handle this in your code to make sure everything works as you intended it to.
node_redis: Warning: Redis server does not require a password, but a password was supplied.
npm ERR! Test failed.  See above for more details.
```
Any insight?

**Comment by PuKoren** [comment]
Hey @jeff-kilbride,

Tests seems to pass fine on Circle, even when we disable the cache.
Did you try to npm update before launching the tests again?

Also, what is the version of your redis server? On Circle it should be [3.0.3](https://circleci.com/docs/build-image-precise/#databases)

**Comment by jeff-kilbride** [comment]
Hi @PuKoren,

Actually, earlier today I deleted my github fork and then re-forked the repo. I had accidentally committed my changes to `develop`, so I wanted to start fresh. I pulled the clean fork, ran `npm install`, and then ran `npm test` in the `develop` branch before creating a new branch with my `zlib` changes. So, the errors above were from a clean pull of the repo and a fresh `npm install`.

I am using the `redis:3.0-alpine` docker image for my tests, which is:
```
/data # redis-server -v
Redis server v=3.0.7 sha=00000000:0 malloc=jemalloc-3.6.0 bits=64 build=93a698f8cd7bf04b
```
I copied my latest `index.js` from my feature branch into my old fork of the project from a little over a week ago, and the tests run fine with that version. Have you tried a clean pull, `npm install`, and `npm test`,  to see if that works?

**Comment by jeff-kilbride** [comment]
@PuKoren,

I submitted my PR a few minutes ago and it's showing all tests passing. However, I am seeing the same error with `npm test` on Circle. This is the output from the `npm test` tab on Circle:
```
npm test

> cache-manager-redis@0.2.3 test /home/ubuntu/node-cache-manager-redis
> jasmine

node_redis: Warning: Redis server does not require a password, but a password was supplied.
node_redis: Deprecated: The SETEX command contains a "undefined" argument.
This is converted to a "undefined" string now and will return an error from v.3.0 on.
Please handle this in your code to make sure everything works as you intended it to.
node_redis: Deprecated: The SETEX command contains a "undefined" argument.
This is converted to a "undefined" string now and will return an error from v.3.0 on.
Please handle this in your code to make sure everything works as you intended it to.
node_redis: Warning: Redis server does not require a password, but a password was supplied.
```
It has an exit code of `0`, which is probably being interpreted by Circle as passing, but it doesn't look like the tests are being run.

**Comment by PuKoren** [comment]
@jeff-kilbride good catch, checking it now

**Comment by PuKoren** [comment]
@jeff-kilbride I created https://github.com/dial-once/node-cache-manager-redis/commit/5eaa089fbd5217007391779dcc84270cc058069b to fix the issue. Seems to be related to Redis 3.x that triggers an error when we try to store an undefined value. Replacing it with `""` make tests pass again on Redis 3.x. The PR is pending review of my mates

Maybe you can cherry-pick the hotfix on your branch for now to make your own tests, or just apply the change manually and not submitting it in your PR

About your old fork, it is possible that this change https://github.com/dial-once/node-cache-manager-redis/commit/c610d2364ca200a57baefcb95002d9e444003e26?diff=unified#diff-458d352f0b6804835faacbee53c3ac7bR81 made the tests fails on Redis 3.x (storing null value)

**Comment by jeff-kilbride** [comment]
@PuKoren  I noticed that Jasmine was updated recently from `2.3.2` to `2.4.1`. Just wondering if that has something to do with it...

**Comment by PuKoren** [comment]
@jeff-kilbride it may be, they tend to often break things

**Comment by jeff-kilbride** [comment]
@PuKoren  Yeah, i just manually updated the `package.json` in my old fork to use Jasmine `2.4.1`, ran `npm update`, and now it's broken...  :\

**Comment by PuKoren** [comment]
Yep the `^` char will take any minor if available, if you want to only update patch-level version you have to use the `~` character

**Comment by jeff-kilbride** [comment]
You're right, my bad. 

I will wait for the merge. I prefer `mocha / chai` to jasmine anyway. FWIW, I downgraded jasmine back to `2.3.2` and the output is working again.

**Comment by PuKoren** [comment]
@jeff-kilbride merged, you should be able to go on 👍

**Comment by jeff-kilbride** [comment]
Mocha tests are working for me!  👍 

I'll add some tests to my PR as soon as I get a chance.

**Comment by jeff-kilbride** [comment]
Resolved by #44.
